File: rust/src/scenes/note_demo/scenarios.rs
//! The note-field animation scenarios the demo scene plays: plain data,
//! one deterministic timeline per rendering behavior worth reviewing.

use crate::nodes::stepfile_player::note_field::HoldVisualState;

pub struct Scenario {
    pub name: String,
    pub notes: Vec<ScenarioNote>,
    pub mines: Vec<ScenarioMine>,
    pub script: Vec<(f64, ScriptAction)>,
    /// `(beat, bpm)` changes; empty means the CLI `--bpm` throughout.
    pub bpms: Vec<(f64, f64)>,
    /// `(beat, seconds)` stops.
    pub stops: Vec<(f64, f64)>,
}

pub struct ScenarioNote {
    pub beat: f64,
    pub column: usize,
    pub quant: u32,
    pub length_beats: Option<f64>,
    pub roll: bool,
}

pub struct ScenarioMine {
    pub beat: f64,
    pub column: usize,
}

/// A scripted stand-in for the gameplay systems that drive the field.
#[derive(Clone, Copy)]
pub enum ScriptAction {
    /// Set the render state of the scenario's i-th note's hold.
    Hold(usize, HoldVisualState),
    /// Apply the hold-OK fade to the i-th note's head.
    Fade(usize),
    /// Vanish the i-th note at the receptor: despawn it and play the arrow
    /// flash on its column, as grading does for taps.
    Vanish(usize),
    /// Press or release a receptor's panel.
    Press(usize, bool),
    /// Blow up the i-th mine.
    ExplodeMine(usize),
}

pub fn scenario_names() -> Vec<String> {
    scenario_matrix()
        .into_iter()
        .map(|scenario| scenario.name)
        .collect()
}

pub fn scenario_matrix() -> Vec<Scenario> {
    use HoldVisualState::{Dropped, Held, Ok, Released};
    use ScriptAction::{ExplodeMine, Fade, Hold, Press, Vanish};

    const QUANTS: [u32; 8] = [4, 8, 12, 16, 24, 32, 48, 64];
    let mut all = Vec::new();
    let mut add = |name: &str,
                   notes: Vec<ScenarioNote>,
                   mines: Vec<ScenarioMine>,
                   script: Vec<(f64, ScriptAction)>| {
        all.push(Scenario {
            name: name.to_string(),
            notes,
            mines,
            script,
            bpms: Vec::new(),
            stops: Vec::new(),
        });
    };

    for quant in QUANTS {
        add(
            &format!("single_quant_{quant}"),
            vec![note(0.0, 1, quant, None)],
            vec![],
            vec![],
        );
    }

    for quant in QUANTS {
        for (label, length) in [
            ("half_beat", 0.5),
            ("one_beat", 1.0),
            ("two_and_a_half_beats", 2.5),
        ] {
            add(
                &format!("hold_quant_{quant}_{label}"),
                vec![note(0.0, 1, quant, Some(length))],
                vec![],
                vec![],
            );
        }
    }

    add(
        "hold_held_to_ok",
        vec![note(0.0, 1, 4, Some(2.0))],
        vec![],
        vec![
            (0.0, Press(1, true)),
            (0.0, Hold(0, Held)),
            (2.0, Hold(0, Ok)),
            (2.0, Fade(0)),
            (2.0, Press(1, false)),
        ],
    );
    add(
        "hold_released_and_regrabbed",
        vec![note(0.0, 1, 4, Some(3.0))],
        vec![],
        vec![
            (0.0, Press(1, true)),
            (0.0, Hold(0, Held)),
            (1.0, Press(1, false)),
            (1.0, Hold(0, Released)),
            (1.75, Press(1, true)),
            (1.75, Hold(0, Held)),
            (3.0, Hold(0, Ok)),
            (3.0, Fade(0)),
            (3.0, Press(1, false)),
        ],
    );
    add(
        "hold_dropped_midway",
        vec![note(0.0, 1, 4, Some(3.0))],
        vec![],
        vec![
            (0.0, Press(1, true)),
            (0.0, Hold(0, Held)),
            (1.0, Press(1, false)),
            (1.0, Hold(0, Released)),
            (1.5, Hold(0, Dropped)),
        ],
    );
    add(
        "hold_head_missed",
        vec![note(0.0, 1, 4, Some(2.0))],
        vec![],
        vec![(0.5, Hold(0, Dropped))],
    );
    add("roll_two_beats", vec![roll(0.0, 1, 4, 2.0)], vec![], vec![]);
    add(
        "roll_held_to_ok",
        vec![roll(0.0, 1, 4, 2.0)],
        vec![],
        vec![
            (0.0, Press(1, true)),
            (0.0, Hold(0, Held)),
            (2.0, Hold(0, Ok)),
            (2.0, Fade(0)),
            (2.0, Press(1, false)),
        ],
    );
    add(
        "hold_chain_one_column",
        vec![
            note(0.0, 1, 4, Some(0.5)),
            note(1.0, 1, 4, Some(0.5)),
            note(2.0, 1, 4, Some(0.5)),
        ],
        vec![],
        vec![],
    );
    add(
        "hold_staircase",
        vec![
            note(0.0, 0, 4, Some(1.0)),
            note(0.5, 1, 8, Some(1.0)),
            note(1.0, 2, 4, Some(1.0)),
            note(1.5, 3, 8, Some(1.0)),
        ],
        vec![],
        vec![],
    );
    add(
        "jump_hold",
        vec![note(0.0, 1, 4, Some(2.0)), note(0.0, 2, 4, Some(2.0))],
        vec![],
        vec![],
    );

    add(
        "tap_vanish_at_receptor",
        vec![note(0.0, 1, 4, None)],
        vec![],
        vec![
            (0.0, Press(1, true)),
            (0.0, Vanish(0)),
            (0.4, Press(1, false)),
        ],
    );
    add(
        "jump",
        vec![note(0.0, 0, 4, None), note(0.0, 3, 4, None)],
        vec![],
        vec![],
    );
    add(
        "every_column",
        vec![
            note(0.0, 0, 4, None),
            note(1.0, 1, 4, None),
            note(2.0, 2, 4, None),
            note(3.0, 3, 4, None),
        ],
        vec![],
        vec![],
    );
    add(
        "stream_16ths",
        (0..16)
            .map(|i| {
                let quant = [4, 16, 8, 16][i % 4];
                note(i as f64 * 0.25, i % 4, quant, None)
            })
            .collect(),
        vec![],
        vec![],
    );

    add("mine", vec![], vec![mine(0.0, 1)], vec![]);
    add(
        "mine_row",
        vec![],
        (0..4).map(|column| mine(0.0, column)).collect(),
        vec![],
    );
    add(
        "mine_exploding",
        vec![],
        vec![mine(0.0, 1)],
        vec![
            (-0.5, Press(1, true)),
            (0.0, ExplodeMine(0)),
            (0.5, Press(1, false)),
        ],
    );

    add(
        "receptors_idle",
        vec![],
        vec![],
        vec![
            (1.0, Press(1, true)),
            (2.0, Press(1, false)),
            (3.0, Press(2, true)),
            (4.0, Press(2, false)),
        ],
    );

    // Tempo gimmicks: under Dynamic speed the spacing per beat must stay
    // uniform while the scroll rate doubles, and a stop must freeze the
    // field; under Constant speed the spacing itself changes instead.
    all.push(Scenario {
        name: "stream_bpm_change".to_string(),
        notes: (0..8).map(|i| note(i as f64, i % 4, 4, None)).collect(),
        mines: vec![],
        script: vec![],
        bpms: vec![(0.0, 125.0), (4.0, 250.0)],
        stops: vec![],
    });
    all.push(Scenario {
        name: "stream_stop".to_string(),
        notes: (0..8).map(|i| note(i as f64, i % 4, 4, None)).collect(),
        mines: vec![],
        script: vec![],
        bpms: vec![],
        stops: vec![(4.0, 1.0)],
    });

    all
}

fn note(beat: f64, column: usize, quant: u32, length_beats: Option<f64>) -> ScenarioNote {
    ScenarioNote {
        beat,
        column,
        quant,
        length_beats,
        roll: false,
    }
}

fn roll(beat: f64, column: usize, quant: u32, length_beats: f64) -> ScenarioNote {
    ScenarioNote {
        roll: true,
        ..note(beat, column, quant, Some(length_beats))
    }
}

fn mine(beat: f64, column: usize) -> ScenarioMine {
    ScenarioMine { beat, column }
}

File: scenes/review/Scenarios.cs

namespace Rhythm;

/// <summary>One note in a scenario: when it appears, which column, and its quantization.</summary>
public record ScenarioNote(double Beat, uint Column, uint Quant, double? LengthBeats, bool Roll);

/// <summary>One mine in a scenario: when and where it appears.</summary>
public record ScenarioMine(double Beat, uint Column);

/// <summary>A scripted action in the demo timeline: advancing the hold state, vanishing notes, etc.</summary>
public abstract record ScriptAction
{
    public sealed record Hold(int Index, HoldVisualState State) : ScriptAction;
    public sealed record Fade(int Index) : ScriptAction;
    public sealed record Vanish(int Index) : ScriptAction;
    public sealed record Press(uint Column, bool Held) : ScriptAction;
    public sealed record ExplodeMine(int Index) : ScriptAction;
}

/// <summary>One animation scenario: plain data describing every rendering behavior.</summary>
public record Scenario(
    string Name,
    IReadOnlyList<ScenarioNote> Notes,
    IReadOnlyList<ScenarioMine> Mines,
    IReadOnlyList<(double Beat, ScriptAction Action)> Script,
    IReadOnlyList<(double Beat, double Bpm)> Bpms,
    IReadOnlyList<(double Beat, double Seconds)> Stops
);

/// <summary>Catalog of all animation scenarios for visual testing and review.</summary>
public static class Scenarios
{
    /// <summary>Gets all scenario names in order.</summary>
    public static IReadOnlyList<string> Names() =>
        Matrix().Select(s => s.Name).ToList();

    /// <summary>Gets all scenarios in order.</summary>
    public static IReadOnlyList<Scenario> Matrix()
    {
        const uint Q4 = 4, Q8 = 8, Q12 = 12, Q16 = 16, Q24 = 24, Q32 = 32, Q48 = 48, Q64 = 64;
        var quants = new[] { Q4, Q8, Q12, Q16, Q24, Q32, Q48, Q64 };

        var scenarios = new List<Scenario>();

        // Single notes at various quantizations
        foreach (var quant in quants)
        {
            scenarios.Add(new(
                $"single_quant_{quant}",
                [new(0.0, 1, quant, null, false)],
                [],
                [],
                [],
                []
            ));
        }

        // Holds at various quantizations and lengths
        foreach (var quant in quants)
        {
            foreach (var (label, length) in new[] {
                ("half_beat", 0.5),
                ("one_beat", 1.0),
                ("two_and_a_half_beats", 2.5)
            })
            {
                scenarios.Add(new(
                    $"hold_quant_{quant}_{label}",
                    [new(0.0, 1, quant, length, false)],
                    [],
                    [],
                    [],
                    []
                ));
            }
        }

        scenarios.Add(new(
            "hold_held_to_ok",
            [new(0.0, 1, 4, 2.0, false)],
            [],
            [
                (0.0, new ScriptAction.Press(1, true)),
                (0.0, new ScriptAction.Hold(0, HoldVisualState.Held)),
                (2.0, new ScriptAction.Hold(0, HoldVisualState.Ok)),
                (2.0, new ScriptAction.Fade(0)),
                (2.0, new ScriptAction.Press(1, false))
            ],
            [],
            []
        ));

        scenarios.Add(new(
            "hold_released_and_regrabbed",
            [new(0.0, 1, 4, 3.0, false)],
            [],
            [
                (0.0, new ScriptAction.Press(1, true)),
                (0.0, new ScriptAction.Hold(0, HoldVisualState.Held)),
                (1.0, new ScriptAction.Press(1, false)),
                (1.0, new ScriptAction.Hold(0, HoldVisualState.Released)),
                (1.75, new ScriptAction.Press(1, true)),
                (1.75, new ScriptAction.Hold(0, HoldVisualState.Held)),
                (3.0, new ScriptAction.Hold(0, HoldVisualState.Ok)),
                (3.0, new ScriptAction.Fade(0)),
                (3.0, new ScriptAction.Press(1, false))
            ],
            [],
            []
        ));

        scenarios.Add(new(
            "hold_dropped_midway",
            [new(0.0, 1, 4, 3.0, false)],
            [],
            [
                (0.0, new ScriptAction.Press(1, true)),
                (0.0, new ScriptAction.Hold(0, HoldVisualState.Held)),
                (1.0, new ScriptAction.Press(1, false)),
                (1.0, new ScriptAction.Hold(0, HoldVisualState.Released)),
                (1.5, new ScriptAction.Hold(0, HoldVisualState.Dropped))
            ],
            [],
            []
        ));

        scenarios.Add(new(
            "hold_head_missed",
            [new(0.0, 1, 4, 2.0, false)],
            [],
            [(0.5, new ScriptAction.Hold(0, HoldVisualState.Dropped))],
            [],
            []
        ));

        scenarios.Add(new(
            "roll_two_beats",
            [new(0.0, 1, 4, 2.0, true)],
            [],
            [],
            [],
            []
        ));

        scenarios.Add(new(
            "roll_held_to_ok",
            [new(0.0, 1, 4, 2.0, true)],
            [],
            [
                (0.0, new ScriptAction.Press(1, true)),
                (0.0, new ScriptAction.Hold(0, HoldVisualState.Held)),
                (2.0, new ScriptAction.Hold(0, HoldVisualState.Ok)),
                (2.0, new ScriptAction.Fade(0)),
                (2.0, new ScriptAction.Press(1, false))
            ],
            [],
            []
        ));

        scenarios.Add(new(
            "hold_chain_one_column",
            [
                new(0.0, 1, 4, 0.5, false),
                new(1.0, 1, 4, 0.5, false),
                new(2.0, 1, 4, 0.5, false)
            ],
            [],
            [],
            [],
            []
        ));

        scenarios.Add(new(
            "hold_staircase",
            [
                new(0.0, 0, 4, 1.0, false),
                new(0.5, 1, 8, 1.0, false),
                new(1.0, 2, 4, 1.0, false),
                new(1.5, 3, 8, 1.0, false)
            ],
            [],
            [],
            [],
            []
        ));

        scenarios.Add(new(
            "jump_hold",
            [
                new(0.0, 1, 4, 2.0, false),
                new(0.0, 2, 4, 2.0, false)
            ],
            [],
            [],
            [],
            []
        ));

        scenarios.Add(new(
            "tap_vanish_at_receptor",
            [new(0.0, 1, 4, null, false)],
            [],
            [
                (0.0, new ScriptAction.Press(1, true)),
                (0.0, new ScriptAction.Vanish(0)),
                (0.4, new ScriptAction.Press(1, false))
            ],
            [],
            []
        ));

        scenarios.Add(new(
            "jump",
            [
                new(0.0, 0, 4, null, false),
                new(0.0, 3, 4, null, false)
            ],
            [],
            [],
            [],
            []
        ));

        scenarios.Add(new(
            "every_column",
            [
                new(0.0, 0, 4, null, false),
                new(1.0, 1, 4, null, false),
                new(2.0, 2, 4, null, false),
                new(3.0, 3, 4, null, false)
            ],
            [],
            [],
            [],
            []
        ));

        scenarios.Add(new(
            "stream_16ths",
            Enumerable.Range(0, 16)
                .Select(i =>
                {
                    var quant = new[] { 4u, 16u, 8u, 16u }[i % 4];
                    return new ScenarioNote(i * 0.25, (uint)(i % 4), quant, null, false);
                })
                .ToList(),
            [],
            [],
            [],
            []
        ));

        scenarios.Add(new(
            "mine",
            [],
            [new(0.0, 1)],
            [],
            [],
            []
        ));

        scenarios.Add(new(
            "mine_row",
            [],
            Enumerable.Range(0, 4)
                .Select(column => new ScenarioMine(0.0, (uint)column))
                .ToList(),
            [],
            [],
            []
        ));

        scenarios.Add(new(
            "mine_exploding",
            [],
            [new(0.0, 1)],
            [
                (-0.5, new ScriptAction.Press(1, true)),
                (0.0, new ScriptAction.ExplodeMine(0)),
                (0.5, new ScriptAction.Press(1, false))
            ],
            [],
            []
        ));

        scenarios.Add(new(
            "receptors_idle",
            [],
            [],
            [
                (1.0, new ScriptAction.Press(1, true)),
                (2.0, new ScriptAction.Press(1, false)),
                (3.0, new ScriptAction.Press(2, true)),
                (4.0, new ScriptAction.Press(2, false))
            ],
            [],
            []
        ));

        scenarios.Add(new(
            "stream_bpm_change",
            Enumerable.Range(0, 8)
                .Select(i => new ScenarioNote(i, (uint)(i % 4), 4, null, false))
                .ToList(),
            [],
            [],
            [(0.0, 125.0), (4.0, 250.0)],
            []
        ));

        scenarios.Add(new(
            "stream_stop",
            Enumerable.Range(0, 8)
                .Select(i => new ScenarioNote(i, (uint)(i % 4), 4, null, false))
                .ToList(),
            [],
            [],
            [],
            [(4.0, 1.0)]
        ));

        return scenarios;
    }
}
